### PR TITLE
MODROLESKC-306 - Implement replacement for dummy capabilities

### PR DESCRIPTION
**Pull request by alekGbuz**
### **Purpose**
[MODROLESKC-306](https://folio-org.atlassian.net/browse/MODROLESKC-306) - Implement replacement for dummy capabilities

### **Approach**
Identify Dummy Capabilities. Dummy capabilities are identified by the dummyCapability boolean field in the CapabilityEntity. 

Modify CapabilityReplacementsService.deduceReplacements to get permissions of retrieves dummy capabilities if their permission names are part of the replacement set.

Modify CapabilityReplacementsService.assignReplacementCapabilities. When a dummy capability is being replaced by a real one, the assignments associated with capability sets which have the dummy capability must be transferred to the new, real capability.

Cleanup of Dummy Capabilities. After the successful migration of assignments, the dummy capability should be deleted. Delete event is published.

---

### **Pre-Review Checklist**

- [x] **Self-reviewed Code** — Reviewed code for issues, unnecessary parts, and overall quality.
- [x] **Change Notes** — NEWS.md updated with clear description and issue key.
- [x] **Testing** — Confirmed changes were tested locally or on dev environment.
- [ ] **Breaking Changes** — Handled all required actions if changes affect API, DB, or interface versions.
  - [ ] API/schema changes
  - [ ] Interface version updates
  - [ ] DB schema changes / migration scripts
- [ ] **New Properties / Environment Variables** — Updated README.md if new configs were added.
- [x] **Environment Recreation Test (if needed)** — Verified that environment can be recreated successfully.

### CODE REVIEW

**Comment by NikitaSedyx** [comment]
# Pull Request Issue Report – PR #258 for mod-roles-keycloak

---

## 1. 🔴 BLOCKING ISSUES (Prevent Merge)

### 1.1 Missing Negative Test Coverage for Error and Failure Scenarios
- **Issue:** No negative, error-path, or rollback tests for dummy capability migration. Only positive/happy path scenarios are covered across critical service and integration test files. Directly undermines confidence in error handling and system integrity.
- **Location:** 
  - `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:56-289`
  - `src/test/java/org/folio/roles/it/DummyCapabilityIT.java:36-149`
  - `src/test/java/org/folio/roles/service/capability/CapabilitySetServiceTest.java` (see file, no negative cases)
- **Impact:** Uncaught bugs or logic failures in migration, assignment, or deletion could lead to orphaned assignments, permissions, or security holes. System recovery on migration failure untested.
- **Required fix:** Implement comprehensive tests covering: failed migrations, partial rollbacks, event emission failures, and malformed input to all migration and replacement logic.

### 1.2 Incomplete Line-Level Traceability and Missing Source/Test Artifacts
- **Issue:** Some core classes noted in reviews (e.g., `CapabilitySetByCapabilitiesUpdater.java`, `CapabilitySetByCapabilitiesUpdaterTest.java`) are not present or not readable in the source; their behavior and test verification cannot be independently reviewed.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java`
  - `src/test/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdaterTest.java`
- **Impact:** Absence of testable source disables full code review and compliance validation, creating a critical risk of invisible bugs or anti-patterns.
- **Required fix:** Ensure all business logic, updaters, and relevant test classes are present in source control and in the PR. If these are auto-generated, provide stubs or references; if missed, add or restore them.

---

## 2. 🚨 SECURITY VULNERABILITIES

### 2.1 Undocumented and Untested Failure/Edge Handling for Assignment Migration/Event Publication
- **Issue:** Absence of tests verifying system security and data integrity on partial or failed assignment transfers leaves risk for incomplete or erroneous permissions, potential privilege escalation or denial-of-access scenarios.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:82-346`
  - `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:56-289`
- **Attack vectors:** Malformed event/data could bypass migration or orphan assignments, especially if event emission or transactional rollback fails.

---

## 3. 🐛 BUGS AND LOGIC ERRORS

### 3.1 Event Emission Verification Missing or Only Mocked
- **Issue:** In capability and set deletion, evidence for event emission is only provided via positive/mock-based tests; no test covers event emission failure, duplicate/deadlock, or idempotency verification.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:306-321`
  - `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:146-256`
- **Impact:** Failure to emit events could cause desynchronization of distributed RBAC state, breaking access control consistency.

### 3.2 Partial Transactional Safety and Rollback Verification
- **Issue:** While transactional annotations are present, real rollback behavior on partial failures (e.g., event publish fails after DB update) is not tested.
- **Location:**
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:82-91, 323-331`
  - `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:146-289`
- **Impact:** Partial state/migration could persist, leaving system in corrupt or inconsistent state.

---

## 4. 📋 REQUIREMENTS FAILURES

### 4.1 Missing/Incomplete Documentation for Complex Migration and Assignment Logic
- **Issue:** No JavaDoc or inline description for non-obvious branches, business rules, or edge cases in migration, deletion, or dummy capability handling.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:151-266, 306-346` (especially on critical branches)
- **Impact:** High maintenance cost, risk of future regressions or incorrect logic changes.

### 4.2 Integration/README/Config Gaps for New or Modified Behavior
- **Issue:** No evidence that README or systems integration documentation have been updated to reflect new migration flows, assignment logic, or required event-driven contract.
- **Location:** 
  - `README.md` (not updated in PR, per check; required for new ENV/config or behavior)
- **Impact:** Onboarding, QA, and deployment-risk elevated; configuration changes or required setups may be missed.

---

## 5. ⚡ PERFORMANCE ISSUES

### 5.1 No Negative/Stress/Scale Test for Large Dummy>Real Assignment Migrations
- **Issue:** No tests or code paths validate assignment migration performance or resource usage when large numbers of capability sets or users are affected.
- **Location:** 
  - `src/test/java/org/folio/roles/it/DummyCapabilityIT.java`, `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java` (missing)
- **Impact:** Potential performance bottlenecks, OOM, or timeouts when bulk events are processed.

---

## 6. 🏗️ ARCHITECTURE & DESIGN PROBLEMS

### 6.1 Reliance on Delegation/Orchestration Without Verifiable Coverage
- **Issue:** Core migration/event emission logic is split across multiple orchestrator/updater services with positive-path-only tests. No cross-service rollback or idempotency test exists.
- **Location:**
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:151-346`
  - `src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java` (missing line-level review)
- **Impact:** Hidden anti-patterns or propagation of bugs across components.

---

## 7. 🧪 TESTING DEFICIENCIES

### 7.1 No Negative/Failure/Edge-Test Coverage in Key Service and Integration Tests
- **Issue:** Only happy path tested; see Blocking Issues above.
- **Location:** 
  - `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:56-289`
  - `src/test/java/org/folio/roles/it/DummyCapabilityIT.java:36-149`
- **Impact:** Uncaught logic, migration, or rollback bugs in production.

### 7.2 Missing Test for Updater Logic
- **Issue:** Test class for `CapabilitySetByCapabilitiesUpdater` is missing or not present for audit; updater logic cannot be validated or confirmed to follow requirements.
- **Location:** 
  - `src/test/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdaterTest.java`
- **Impact:** Critical functionality unsafely untested.

---

## 8. 📝 DOCUMENTATION GAPS

### 8.1 No JavaDoc for Complex Service & Orchestrator Methods
- **Issue:** No or minimal JavaDoc comments for critical logic in assignment migration, deletion, and event emission branches.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java` (entire file, especially `assignReplacementCapabilities`, `unassignReplacedCapabilities`)
- **Impact:** Reviewer and maintainer uncertainty, future bug risk.

### 8.2 No Edge Case or Failure Documentation for Deletion, Assignment, or Event Emission
- **Issue:** Expected side effects, edge behaviors, and error conditions not described anywhere in code or docs.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java`, `README.md`
- **Impact:** System integration failures likely under real-world conditions.

---

## 9. 🔧 CODE QUALITY ISSUES

### 9.1 Maintainability/Readability Problems in Large Service Classes
- **Issue:** Complex logic in large methods (e.g., `processReplacements`, `assignReplacementCapabilities`, `replaceLoadable`, `unassignReplacedCapabilities`) without clear breakdown, excessive local branching, or inline lambda use without doc.
- **Location:** 
  - `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:82-346`
- **Impact:** Hard to reason about, high future bug risk.

---

## 10. 🔗 INTEGRATION PROBLEMS

### 10.1 No Regression/Compatibility/Contract Test for Upstream/Downstream Invocation
- **Issue:** Changes in migration event handling are not fully documented or validated for upstream/downstream API/data contract compatibility; system integration is tested only for positive path.
- **Location:**
  - `src/test/java/org/folio/roles/it/DummyCapabilityIT.java:36-149`
- **Impact:** Integration break risk if contract changes or if upstream/downstream services evolve.

---

## 11. ⚙️ CONFIGURATION RISKS

### 11.1 Missing README/Deployment Docs for Changed Behavior or Configs
- **Issue:** Any new event types, Kafka topics, or migration procedures are not documented for deployment teams.
- **Location:**
  - `README.md`, deployment docs (not present in PR)
- **Impact:** Production or test instability on deployment.

---

## 12. **Issue Priority Matrix**

| Issue Category   | Count | Critical | High | Medium | Example Location                                              |
|------------------|-------|----------|------|--------|--------------------------------------------------------------|
| Security         | 1     | 1        | 0    | 0      | src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:82-346 |
| Bugs             | 3     | 1        | 2    | 0      | src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:306-321 |
| Performance      | 1     | 0        | 1    | 0      | src/test/java/org/folio/roles/it/DummyCapabilityIT.java      |
| Architecture     | 1     | 0        | 1    | 0      | src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java |
| Testing          | 2     | 2        | 0    | 0      | src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:56-289 |
| Documentation    | 2     | 0        | 2    | 0      | src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:151-266 |
| Integration      | 1     | 0        | 1    | 0      | src/test/java/org/folio/roles/it/DummyCapabilityIT.java:36-149 |
| Configuration    | 1     | 0        | 1    | 0      | README.md                                                    |

---

## 13. **Blocking Issues List**

### **🔴 Critical Blockers**
- **Missing negative/edge/failure tests for dummy capability migration and deletion**  
  - Location: `src/test/java/org/folio/roles/service/capability/CapabilityReplacementsServiceTest.java:56-289`, `src/test/java/org/folio/roles/it/DummyCapabilityIT.java:36-149`  
  - Impact: Defects in migration/event handling will remain undetected, risking data and permission integrity.  
  - Required fix: Add coverage for error/failure, rollback, malformed event/data, and event emission failure paths.

- **Invisible code pieces: missing source/test for CapabilitySetByCapabilitiesUpdater**  
  - Location: `src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java`, `src/test/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdaterTest.java`  
  - Impact: Core migration logic cannot be verified for correctness or test coverage.

### **🔴 Severe Issues**
- **Lack of transactional rollback/error-path test coverage**  
  - Location: `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:323-331`, related tests  
  - Impact: Partial/inconsistent state risk if event emission or DB write fails.

- **Absence of documentation on complex migration/event emission flow**  
  - Location: `src/main/java/org/folio/roles/service/capability/CapabilityReplacementsService.java:151-266`, `README.md`  
  - Impact: Risk for regressions, misunderstandings, and incorrect bug/security fixes in future releases.

### **🟠 Major Problems**
- **No integration, performance, or scale tests for bulk migration**  
  - Location: `src/test/java/org/folio/roles/it/DummyCapabilityIT.java`  
- **No regression/contract validation for event contract**  
  - Location: `src/test/java/org/folio/roles/it/DummyCapabilityIT.java`  
- **Undocumented or untested event emission failures, edge/corner-cases, and transactional side effects**

---

## Final Risk Assessment

- **Overall Risk Level:** ❌ HIGH RISK (Critical blockers and severe omissions in test and documentation)
- **Blocking Issues Count:** 2 critical, 2 severe, several major
- **Requirements Coverage:** Multiple major requirements (negative/error test, rollback, doc) missing/incomplete
- **Security Risk:** CRITICAL (potential privilege or denial-of-access bugs)
- **Bug Density:** Elevated due to untested negative paths (estimate >1/50 LOC in affected core services)
- **Test Coverage Gap:** >30% untested code in critical new logic (all edge/error/rollback paths)
- **Must-Fix Count:** 6+ before merge

---

## Additional Issue Categories Identified
- **None observed for Accessibility/Internationalization/Concurrency/Resource leaks from provided material; review focused on business/service/migration logic per requirements.**

---

### Key Action Items

1. **Implement negative and failure-path tests** for all dummy migration logic, assignment transfer, event emission (including rollback and incomplete state).
2. **Ensure all source and test files are present & reviewable**, especially orchestrators/updaters.
3. **Refactor and document complex service logic** with full JavaDoc in every method handling migration, event publication, and deletion, especially all multi-branch flows.
4. **Extend and document configuration/README** to reflect changed requirements, new capabilities, and integration contracts.
5. **Verify and enforce transactional integrity** and state rollback on all DB/messaging failures.
6. **Provide regression and validation tests for event contract** and system integration.

---

**Do NOT merge this PR until all of the above blocking and severe issues are fully addressed.**

File: src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  /**
+   * Update capability set by capabilities and publish update event.
+   *
+   * @param capabilitiesToAdd - capabilities to add to capability set
+   * @param capabilitySet - capability set to update
+   */
+  public void update(CapabilitySet capabilitySet, List<Capability> capabilitiesToAdd) {
+    log.info("Update capability set by capabilities : capability set name - {},  capabilities - {}",
+      capabilitySet.getName(),
+      toStream(capabilitiesToAdd).map(Capability::getName).collect(Collectors.joining(", ")));
```
[the remark this answers is not in the document]
Comment (alekGbuz) [reply src/main/java/org/folio/roles/service/capability/CapabilitySetByCapabilitiesUpdater.java]: updated
